### ISSUE TITLE
Temp file creation on Windows name conflict/race condition

### ISSUE THREAD

**Issue by aggsol** [opencv/opencv#19648]
- OpenCV => 2.4.3 and later
- Operating System / Platform => Windows 
- Compiler => any

##### Detailed description

For context see [core/src/system.cpp](https://github.com/opencv/opencv/blob/f70e80a6bacec67842757ffbed7b9a643488388d/modules/core/src/system.cpp#L982)

The function cv::tempfile() uses [::GetTempFileNameA](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-gettempfilenamea) to create a temp file. It immediately deletes it (see next line) instead of reserving it and only returns the filename. Other OpenCV process at the same time can create a temp files with same name resulting in a name collision.

MS recommends in the remarks of using a GUID when many of temp files are required.

Proposed solution: Use the already implemented (GetFileNameWinRT)[https://github.com/opencv/opencv/blob/f70e80a6bacec67842757ffbed7b9a643488388d/modules/core/src/system.cpp#L198] on line 198, system.cpp

**Comment by saskatchewancatch** [comment]
Makes sense. `CoCreateGuid`CoCreateGuid` itself is available in Win32 API space (i.e. it would make sense within the #ifdef scope). Although running git blame on the call to `DeleteFileA` on L985 shows it was inserted there to fix an issue from 9 years ago (issue report is not available as it predates transition from SVN to Github), curious why it'd be there other than to ensure the caller gets assurance that the temp file doesn't already exist

**Comment by aggsol** [comment]
@saskatchewancatch As long as there is no possible name collision anymore it won't matter if the file exists for the caller.
